### ISSUE TITLE
Update container if attributes change

### ISSUE THREAD

**Issue by dpkp** [bflad/chef-docker#309]
It would be useful to "redeploy" a container if any key attributes change:

for example, when working with containers that use release tags, it would be useful to do this:
```
docker_container 'foobar' do
  action :run
  image "foobar:#{tag}"
  detach true
end
```
and have chef converge the container to the correct image if the value of tag is changed (via attribute etc). the current behavior, as I understand it, is that changing the value of `tag` will not update a running container.

**Comment by jagalan** [comment]
Agree that could be useful to have that. For now it can be solved by using the tag in docker_pull and then notifying docker_container with redeploy.

**Comment by dpkp** [comment]
Hi jagalan -- unfortunately docker_pull + redeploy only works if the pulled image:tag changes . If, for example, the images are already available locally then chef does not mark the docker_image resource as changed, and so the :redeploy notification does not get sent to the container. So although redeploy works sort-of, it's not really a suitable work around.

**Comment by wpietri** [comment]
A related issue I ran into is redeploying after I change a volume mapping. As far as I could tell, nothing happened when I added a new mapping, so I manually did `docker rm -f` and ran the chef client again. Not a big problem, but I was puzzled.

**Comment by wpietri** [comment]
I'd like to solve this one, and I'll have some time to work on it the week of 8/3. If anybody has suggestions, I'm all ears.

**Comment by dpkp** [comment]
my PR (#310) should fix. take a look and let me know if you have comments

**Comment by yonah-codefresh** [comment]
I had a similar problem. Redeploy doesn't discriminate between containers that had/had not been changed. When the change was in a dependency like the image pulled, this could be solved with notifications but that meant using a nothing action by default. With a default of nothing, the containers wouldn't pick up changes to their own attributes.
What makes it harder to solve is there was no real code to get the parameters of a running container and I was too lazy to write it ;) Maybe now that the cookbook is speaking to the API it will be easier.
To solve it, I hashed new_resource.to_json on container start and saved to a file, then checked if the hashes matched before redeploy. If the hashes didn't match, I replaced the old hash and let the redeploy do it's work. It is still buggy so I'm sure there is a better way but maybe that helps.

**Comment by someara** [comment]
Just a heads up.. I've started working on this. The state comparison is going to be tricky because there are various fields where the input and output end up in different formats. (ulimits, for example.)

**Comment by jperville** [comment]
My workaround for this issue is to store the creation parameters in a file (from a wrapper LWRP), then notify the container's redeploy action when the file contents change. This allows for redeploying when bind mounts, ports, restart policy or whatever important parameter is changed.

**Comment by someara** [comment]
Implemented a load_current_resource in 769554333e5175e181c63c945086b5f8942f1ed0 . Released as 1.0.15

**Comment by chasebolt** [comment]
I think we need to reopen this issue.

I am currently testing with the linuxserver/nzbget on docker hub.

Docker::Container.get() is returning information about the docker container that contains what we are passing into it with docker_container and what the Dockerfile has included. This causes resource_changes to always return changes and docker containers recreated on every chef run; I use action :run on all containers.

Since we can't rely on the container info we get back from the docker api to match exactly what we are creating them with. The only two options I see on how to properly deal with this is either we dump new_resource into a file and redeploy on changes, or we set new_resource into a saved node attribute (though this wouldnt work in a chef-solo run). Personally I think both of these options aren't great, but atm I can't think of any other way.

```ruby
docker_image 'linuxserver/nzbget' do
  notifies :redeploy, 'docker_container[nzbget]', :immediately
end

docker_container 'nzbget' do
  repo 'linuxserver/nzbget'
  port '6789:6789'
  env [
    'PUID=10000',
    'PGID=10000'
  ]
  binds [
    "#{node['storage']['hdd']}/apps/nzbget:/config",
    "#{node['storage']['hdd']}/downloads:/downloads"
  ]
  action :run
end
```
```
[2015-09-15T11:44:35-07:00] DEBUG: Provider for action run on resource docker_container[nzbget] is Chef::Provider::DockerContainer
[2015-09-15T11:44:35-07:00] DEBUG: DOCKER: command - current:["/sbin/my_init"]: parsed:[]:
[2015-09-15T11:44:35-07:00] DEBUG: DOCKER: user - current:: new::
[2015-09-15T11:44:35-07:00] DEBUG: DOCKER: env - current:["PUID=10000", "PGID=10000", "PATH=/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin", "DEBIAN_FRONTEND=noninteractive", "HOME=/root", "TERM=screen"]: parsed:["PUID=10000", "PGID=10000"]:
[2015-09-15T11:44:35-07:00] DEBUG: DOCKER: entrypoint - current:: parsed::
[2015-09-15T11:44:35-07:00] DEBUG: DOCKER: volumes - current:{"/config"=>{}, "/downloads"=>{}}: parsed::
[2015-09-15T11:44:35-07:00] DEBUG: DOCKER: log_config - current:{"Type"=>"json-file", "Config"=>{}}: serialized:{"Type"=>"json-file", "Config"=>{}}:
[2015-09-15T11:44:35-07:00] DEBUG: DOCKER: working_dir - current:: new::
[2015-09-15T11:44:35-07:00] DEBUG: DOCKER: ulimits - current:: new::
[2015-09-15T11:44:35-07:00] DEBUG: DOCKER: change - :exposed_ports
[2015-09-15T11:44:35-07:00] DEBUG: DOCKER: change - :volumes
```
